src/Database/Connection.php

```diff
@@ -183,24 +186,43 @@ public function configName(): string
      * @throws \Cake\Database\Exception\MissingDriverException When a driver class is missing.
      * @throws \Cake\Database\Exception\MissingExtensionException When a driver's PHP extension is missing.
      * @return $this
+     * @deprecated 4.4.0 Setting the driver is deprecated. Use the connection config instead.
      */
     public function setDriver($driver, $config = [])
     {
+        deprecationWarning('Setting the driver is deprecated. Use the connection config instead.');
+
+        $this->_driver = $this->createDriver($driver, $config);
+
+        return $this;
+    }
+
+    /**
+     * Creates driver from name, class name or instance.
+     *
+     * @param \Cake\Database\DriverInterface|string $name Driver name, class name or instance.
+     * @param array $config Driver config if $name is not an instance.
+     * @return \Cake\Database\DriverInterface
+     * @throws \Cake\Database\Exception\MissingDriverException When a driver class is missing.
+     * @throws \Cake\Database\Exception\MissingExtensionException When a driver's PHP extension is missing.
+     */
+    protected function createDriver($name, array $config): DriverInterface
+    {
+        $driver = $name;
         if (is_string($driver)) {
             /** @psalm-var class-string<\Cake\Database\DriverInterface>|null $className */
             $className = App::className($driver, 'Database/Driver');
             if ($className === null) {
-                throw new MissingDriverException(['driver' => $driver]);
+                throw new MissingDriverException(['driver' => $driver, 'connection' => $this->configName()]);
             }
             $driver = new $className($config);
         }
+
         if (!$driver->enabled()) {
```

Comment:
This throws a php error if the driver isn't an instance of DriverInterface. I would prefer a TypeError thrown when trying to return the value instead.

We could move this enabled check up to the constructor.